File: scraper.py
import asyncio
import aiohttp
import random
import urllib.parse
from bs4 import BeautifulSoup
import os
from typing import List, Dict, Any, Callable

try:
    import google.generativeai as genai
    HAS_GENAI = True
except ImportError:
    HAS_GENAI = False

USER_AGENTS = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0"
]

async def check_robots_txt(url: str, user_agent: str = "SearchScrapeBot") -> bool:
    try:
        parsed = urllib.parse.urlparse(url)
        robots_url = f"{parsed.scheme}://{parsed.netloc}/robots.txt"
        
        async with aiohttp.ClientSession() as session:
            async with session.get(robots_url, headers={"User-Agent": random.choice(USER_AGENTS)}, timeout=5) as res:
                if res.status != 200:
                    return True # Default allowed if missing or auth error
                    
                text = await res.text()
                lines = text.split("\n")
                
                applies_to_us = False
                disallowed_paths = []
                
                for line in lines:
                    line = line.strip()
                    if line.startswith("#") or not line:
                        continue
                    if ":" not in line:
                        continue
                    
                    directive, val = line.split(":", 1)
                    directive = directive.strip().lower()
                    val = val.strip()
                    
                    if directive == "user-agent":
                        ua = val.lower()
                        if ua == "*" or user_agent.lower() in ua:
                            applies_to_us = True
                        else:
                            applies_to_us = False
                            
                    if applies_to_us and directive == "disallow":
                        if val:
                            disallowed_paths.append(val)
                            
                path_to_check = parsed.path or "/"
                for dis in disallowed_paths:
                    # Quick prefix match or match all
                    if dis == "/" or path_to_check.startswith(dis):
                        return False
    except:
        pass
    return True

gemini_lock = asyncio.Lock()

async def generate_gemini_summary(text_corpus: str, keyword: str) -> str:
    gemini_key = os.environ.get("GEMINI_API_KEY")
    if not gemini_key or not HAS_GENAI:
        return ""
        
    async with gemini_lock:
        # Rate limit safety delay
        await asyncio.sleep(3.0)
        
        models_to_try = [
            "gemini-2.5-flash",
            "gemini-2.5-pro"
        ]
        for model_name in models_to_try:
            attempts = 2
            for attempt in range(1, attempts + 1):
                try:
                    genai.configure(apiKey=gemini_key)
                    model = genai.GenerativeModel(model_name)
                    
                    prompt = f"""Summarize this web page text in context of target keyword: "{keyword}".
Focus on concrete details, keep it under 100 words. Return only the summary text.
Web text:
{text_corpus[:3000]}"""
                    
                    # Run model invocation
                    response = model.generate_content(prompt)
                    if response and response.text:
                        return response.text.strip()
                except Exception as e:
                    err_str = str(e).lower()
                    is_rate_limit = "429" in err_str or "resource" in err_str or "quota" in err_str
                    is_overloaded = "503" in err_str or "overloaded" in err_str or "unavailable" in err_str
                    
                    is_quoted_or_scary = "quota" in err_str or "exhausted" in err_str or "limit" in err_str or "rate" in err_str or "429" in err_str or "503" in err_str
                    display_err = "Temporary API limitation or service demand spike." if is_quoted_or_scary else (str(e)[:80] + "..." if len(str(e)) > 80 else str(e))
                    print(f"[Gemini Info] Scraper summary attempt {attempt}/{attempts} for {model_name} request unsuccessful: {display_err}")
                    
                    if is_rate_limit or is_overloaded:
                        print(f"[Gemini Info] Overload or Rate limit on {model_name}. Advancing directly to alternative fallback model.")
                        break # Break current attempts to try next model immediately
                        
                    if attempt < attempts:
                        await asyncio.sleep(2.0)
        return ""

async def scrape_single_page(url: str, keyword: str, delay_min: float, delay_max: float, run_ai_summaries: bool) -> Dict[str, Any]:
    # 1. Robots analysis
    allowed = await check_robots_txt(url)
    if not allowed:
        return {"status": "failed", "error_reason": "Blocked by domain robots.txt directives"}
        
    # 2. Polite Delay
    if delay_min > 0:
        wait_time = random.uniform(delay_min, delay_max)
        await asyncio.sleep(wait_time)
        
    # 3. Request page
    headers = {
        "User-Agent": random.choice(USER_AGENTS),
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Referer": "https://www.google.fr/"
    }
    
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(url, headers=headers, timeout=15) as res:
                if res.status != 200:
                    return {"status": "failed", "error_reason": f"HTTP response status code: {res.status}"}
                    
                html = await res.text()
                soup = BeautifulSoup(html, "html.parser")
                
                # strip clutter tags
                for script in soup(["script", "style", "nav", "footer", "header", "noscript", "iframe"]):
                    script.extract()
                    
                title = soup.title.string.strip() if soup.title else soup.h1.get_text(strip=True) if soup.h1 else "Untitled Document"
                
                # Meta description
                meta_desc = ""
                meta_tag = soup.find("meta", attrs={"name": "description"})
                if meta_tag and meta_tag.get("content"):
                    meta_desc = meta_tag["content"].strip()
                    
                # Images
                images = []
                for img in soup.find_all("img"):
                    src = img.get("src")
                    if src and src.startswith("http") and len(images) < 5:
                        images.append(src)
                        
                # Rebuild main body paragraphs
                paragraphs = []
                for tag in soup.find_all(["p", "h1", "h2", "h3"]):
                    txt = tag.get_text(strip=True)
                    if len(txt) > 30 and "{" not in txt and "}" not in txt:
                        paragraphs.append(txt)
                        
                full_content = "\n\n".join(paragraphs)
                if not paragraphs:
                    return {"status": "failed", "error_reason": "No readable textual paragraphs extracted from url body"}
                    
                # Create summaries
                summary = ""
                if run_ai_summaries:
                    summary = await generate_gemini_summary(full_content, keyword)
                    
                if not summary:
                    summary = " ".join(paragraphs[:3])
                    if len(summary) > 300:
                        summary = summary[:297] + "..."
                        
                return {
                    "status": "success",
                    "title": title,
                    "description": meta_desc or f"Web resource discussing {keyword}",
                    "images": images,
                    "summary": summary,
                    "content": full_content[:6000]
                }
    except asyncio.TimeoutError:
        return {"status": "failed", "error_reason": "Request timed out after 15 seconds"}
    except Exception as e:
        return {"status": "failed", "error_reason": str(e)}

async def crawl_concurrent_batch(
    results_list: List[Dict[str, Any]], 
    keyword: str, 
    concurrency_limit: int, 
    delay_min: float, 
    delay_max: float,
    run_ai_summaries: bool,
    on_completed_callback: Callable[[int, Dict[str, Any]], None]
) -> List[Dict[str, Any]]:
    
    semaphore = asyncio.Semaphore(concurrency_limit)
    finalized = [None] * len(results_list)
    
    async def worker(idx: int, item: Dict[str, Any]):
        async with semaphore:
            # Mark active
            on_completed_callback(idx, {**item, "status": "scraping"})
            
            updates = await scrape_single_page(
                item["url"], 
                keyword, 
                delay_min, 
                delay_max, 
                run_ai_summaries
            )
            
            final_item = {**item, **updates}
            finalized[idx] = final_item
            on_completed_callback(idx, final_item)
            
    tasks = [worker(i, item) for i, item in enumerate(results_list)]
    await asyncio.gather(*tasks)
    return finalized

File: server/scraper.ts
import axios from "axios";
import * as cheerio from "cheerio";
import { generateContentWithRetry } from "./gemini_client";
import { AppSettings, WebpageResult } from "./database";

// Modern User-Agents for request spoofing
const AGENTS = [
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
  "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
  "Mozilla/5.0 (iPhone; CPU iPhone OS 17_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.2 Mobile/15E148 Safari/605.1.15"
];

function getRandomAgent(): string {
  return AGENTS[Math.floor(Math.random() * AGENTS.length)];
}

/**
 * Parses a simple robots.txt of a domain to enforce ethical scraping
 */
export async function isAllowedByRobotsTxt(url: string, userAgent = "SearchScrapeBot"): Promise<{ allowed: boolean; reason?: string }> {
  try {
    const parsedUrl = new URL(url);
    const robotsUrl = `${parsedUrl.origin}/robots.txt`;
    
    const response = await axios.get(robotsUrl, {
      headers: { "User-Agent": getRandomAgent() },
      timeout: process.env.VERCEL ? 1500 : 5000,
      validateStatus: (stat) => stat === 200
    });

    const lines = response.data.split(/\r?\n/);
    let appliesToUs = false;
    const disallows: string[] = [];

    for (const line of lines) {
      const trimmed = line.trim();
      if (trimmed.startsWith("#") || trimmed === "") continue;

      const partSplit = trimmed.split(":");
      if (partSplit.length < 2) continue;

      const directive = partSplit[0].toLowerCase().trim();
      const value = partSplit.slice(1).join(":").trim();

      if (directive === "user-agent") {
        const ua = value.toLowerCase();
        if (ua === "*" || ua.includes(userAgent.toLowerCase())) {
          appliesToUs = true;
        } else {
          appliesToUs = false;
        }
      }

      if (appliesToUs && directive === "disallow") {
        if (value) {
          disallows.push(value);
        }
      }
    }

    // Check if path is matched in disallows
    const currentPath = parsedUrl.pathname + parsedUrl.search;
    for (const pattern of disallows) {
      // Escape special regex chars except *
      const regexStr = "^" + pattern
        .replace(/[-\/\\^$*+?.()|[\]{}]/g, "\\$&")
        .replace(/\\\*/g, ".*");
      const regex = new RegExp(regexStr);
      if (regex.test(currentPath) || (pattern === "/" && currentPath === "/")) {
        return { allowed: false, reason: `Blocked by domain's robots.txt rule: 'Disallow: ${pattern}'` };
      }
    }

    return { allowed: true };
  } catch (err: any) {
    // If robots.txt doesn't exist or times out, we assume standard accessibility (liberal fallback)
    return { allowed: true };
  }
}

/**
 * Creates summaries utilizing Gemini server-side SDK with throttling and retries
 */
async function generateAiSummary(rawContent: string, keyword: string): Promise<string> {
  const truncated = rawContent.slice(0, 3000); // Feed a reasonable chunk
  const prompt = `Analyze this scraped text content from a web page and summarize it in relation to the query keyword: "${keyword}".
Focus on what the site offers, its perspective, and key information concerning the keyword. Keep the summary under 120 words.
Make it professional, objective, and dense with details.
Scraped web page text:
${truncated}`;

  try {
    const timeoutPromise = new Promise<never>((_, reject) =>
      setTimeout(() => reject(new Error("Gemini summary generation timeout")), 3000)
    );

    const fetchPromise = generateContentWithRetry(prompt);
    const text = await Promise.race([fetchPromise, timeoutPromise]);
    return text || "";
  } catch (e: any) {
    console.warn("All AI summarization attempts failed, falling back to local extractor:", e.message || e);
    return "";
  }
}

/**
 * Main webpage scraper routine
 */
export async function scrapePage(
  result: WebpageResult, 
  keyword: string, 
  settings: AppSettings
): Promise<Partial<WebpageResult>> {
  try {
    // 1. Robots.txt ethical boundary checking
    const robotStatus = await isAllowedByRobotsTxt(result.url);
    if (!robotStatus.allowed) {
      return {
        status: "failed",
        errorReason: robotStatus.reason || "Disallowed by robots.txt directory permissions"
      };
    }

    // 2. Polite Delay implementation (to prevent overloading targets)
    if (settings.politeModeDelayMin > 0) {
      const waitSec = settings.politeModeDelayMin + Math.random() * (settings.politeModeDelayMax - settings.politeModeDelayMin);
      await new Promise(resolve => setTimeout(resolve, waitSec * 1000));
    }

    // 3. Page fetch
    const response = await axios.get(result.url, {
      headers: {
        "User-Agent": getRandomAgent(),
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webkit,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.5",
        "Referer": "https://www.google.com/"
      },
      timeout: process.env.VERCEL ? 3500 : 15000, // 15s absolute timeout
      maxRedirects: 3
    });

    const html = response.data;
    if (typeof html !== "string") {
      throw new Error("Returned invalid content-type, expected HTML document text");
    }

    const $ = cheerio.load(html);

    // Clean page: strip headers, side nav boilerplates, styles, and scripts
    $("script, style, noscript, iframe, svg, nav, footer, header, #header, #footer, .sidebar, .nav, .menu, ad, advertisement").remove();

    // Grab headers to update page text title
    let pageTitle = $("title").text().trim();
    if (!pageTitle) {
      pageTitle = $("h1").first().text().trim() || result.title || "Untitled Document";
    }

    const metaDescription = $('meta[name="description"]').attr("content")?.trim() || result.description;

    // Collect images
    const images: string[] = [];
    $("img").each((_, img) => {
      const src = $(img).attr("src");
      if (src && !src.startsWith("data:") && (src.startsWith("http://") || src.startsWith("https://") || src.startsWith("/"))) {
        try {
          const absoluteUrl = src.startsWith("/") ? new URL(src, result.url).href : src;
          if (images.length < 5) {
            images.push(absoluteUrl);
          }
        } catch (_) {}
      }
    });

    // Reconstruct structural main body text
    const paragraphs: string[] = [];
    $("h1, h2, h3, p").each((_, elem) => {
      const txt = $(elem).text().trim();
      // Only keep readable sentences
      if (txt.length > 25 && !txt.includes("{") && !txt.includes("}")) {
        paragraphs.push(txt);
      }
    });

    const fullContent = paragraphs.join("\n\n");
    if (fullContent.length < 100) {
      throw new Error("Page contains insufficient readable content or represents a boilerplate redirection layer");
    }

    // Create summary
    let summaryText = "";
    if (settings.enableGeminiSummarization) {
      summaryText = await generateAiSummary(fullContent, keyword);
    }

    // Default structural fallback summary if Gemini wasn't used or failed
    if (!summaryText) {
      summaryText = paragraphs.slice(0, 3).join(" ") || fullContent.slice(0, 300);
      if (summaryText.length > 400) {
        summaryText = summaryText.slice(0, 397) + "...";
      }
    }

    return {
      title: pageTitle,
      description: metaDescription,
      summary: summaryText,
      content: fullContent.slice(0, 8000), // Cap database storage size safely
      images: images,
      status: "success"
    };

  } catch (err: any) {
    let reason = err.message || "Unknown retrieval error";
    if (err.response) {
      reason = `HTTP Error Code: ${err.response.status} (${err.response.statusText || "unauthorized"})`;
    } else if (err.code === "ECONNABORTED") {
      reason = "Scraping timed out after 15 seconds";
    }
    return {
      status: "failed",
      errorReason: reason
    };
  }
}

/**
 * Runs a concurrent batch of page scrapes with pool restriction
 */
export async function scrapeBatch(
  results: WebpageResult[], 
  keyword: string, 
  settings: AppSettings,
  onItemCompleted: (index: number, updatedItem: WebpageResult) => void
): Promise<WebpageResult[]> {
  const activeLimit = Math.max(1, Math.min(30, settings.concurrencyLimit));
  const workList = [...results];
  const finishedList: WebpageResult[] = [];
  
  // Custom pooled concurrency resolver
  const poolLauncher = async () => {
    while (workList.length > 0) {
      const idx = results.length - workList.length;
      const item = workList.shift();
      if (!item) break;

      // Update status to scraping
      const scrapingItem = { ...item, status: "scraping" as const };
      onItemCompleted(idx, scrapingItem);

      // Fetch
      const updates = await scrapePage(scrapingItem, keyword, settings);
      const finalized = { ...scrapingItem, ...updates };
      
      finishedList.push(finalized);
      onItemCompleted(idx, finalized);
    }
  };

  const poolThreads = Array(activeLimit).fill(null).map(() => poolLauncher());
  await Promise.all(poolThreads);

  return results.map(orig => finishedList.find(f => f.url === orig.url) || { ...orig, status: "failed" as const, errorReason: "Queue cancelled" });
}
